CoseSignTool/CoseSign1/Extensions/CoseSign1MessageDetachedSignatureExtensions.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace CoseSign1.Extensions;
+
+/// <summary>
+/// Class which extends <see cref="CoseSign1Message"/> for detached signature use cases.
+/// </summary>
+/// <remarks>
+/// Logging is done through Trace.TraceError and Debug.WriteLine.
+/// </remarks>
+public static class CoseSign1MessageDetachedSignatureExtensions
+{
+    private static readonly string AlgorithmGroupName = "algorithm";
+    private static readonly Regex HashMimeTypeExtension = new (@$"(?<extension>\+hash-(?<{AlgorithmGroupName}>\w+))", RegexOptions.Compiled);
+
+    /// <summary>
+    /// Extracts the hash name from the hash extension within the mime type
+    /// </summary>
+    /// <param name="this">The COSE Sign1 Message to evaluate</param>
+    /// <param name="name">The discovered Hash Algorithm Name from the Content Type Protected Header value of the CoseSign1Message.</param>
+    /// <returns>True if successful in extracting a HashAlgorithmName from the Content Type Protected Header; False otherwise.</returns>
+    public static bool TryGetDetachedSignatureAlgorithm(this CoseSign1Message @this, out HashAlgorithmName name)
+    {
+        if(@this == null)
+        {
+            Trace.TraceError($"{nameof(TryGetDetachedSignatureAlgorithm)} was called on a null CoseSign1Message object");
+            return false;
+        }
+
+        if(!@this.ProtectedHeaders.ContainsKey(CoseHeaderLabel.ContentType))
+        {
+            Trace.TraceError($"{nameof(TryGetDetachedSignatureAlgorithm)} was called on a CoseSign1Message object({@this.GetHashCode()}) without the ContentType protected header present.");
+            return false;
+        }
+
+        string contentType = @this.ProtectedHeaders[CoseHeaderLabel.ContentType].GetValueAsString();
+        if(string.IsNullOrEmpty(contentType))
+        {
+            Trace.TraceError($"{nameof(TryGetDetachedSignatureAlgorithm)} was called on a CoseSign1Message object({@this.GetHashCode()}) without the ContentType protected header being a string value.");
+            return false;
+        }
+
+        Match mimeMatch = HashMimeTypeExtension.Match(contentType);
+        if(!mimeMatch.Success)
+        {
+            Trace.TraceError($"{nameof(TryGetDetachedSignatureAlgorithm)} was called on a CoseSign1Message object({@this.GetHashCode()}) with the ContentType protected header being \"{contentType}\" however it did not match the regex pattern \"{HashMimeTypeExtension}\".");
+            return false;
+        }
+
+        name = new HashAlgorithmName(mimeMatch.Groups[AlgorithmGroupName].Value.ToUpperInvariant());
+        Debug.WriteLine($"{nameof(TryGetDetachedSignatureAlgorithm)} extracted hash algorithm name: {name.Name}, returning true");
+        return true;
+    }
+
+    /// <summary>
+    /// Can be used to determine if a given CoseSign1Message is encoded detached signature for an artifact.

```

Comment:
Standard text (MSDN style) would be
"Determines whether the current CoseSign1Message includes a detached signature."